File: sdk/typescript/src/keypair.ts
/**
 * Keypair management for Aether
 * 
 * Ed25519 keypair generation and signing
 */

import * as ed25519 from '@noble/ed25519';
import { sha256 } from '@noble/hashes/sha256';
import { Keypair, Bytes, Address, Signature } from './types';

export class AetherKeypair implements Keypair {
  public readonly publicKey: Bytes;
  public readonly secretKey: Bytes;
  public readonly address: Address;

  private constructor(secretKey: Bytes, publicKey: Bytes) {
    this.secretKey = secretKey;
    this.publicKey = publicKey;
    this.address = AetherKeypair.publicKeyToAddress(publicKey);
  }

  /**
   * Generate a new random keypair
   */
  static async generate(): Promise<AetherKeypair> {
    const secretKey = ed25519.utils.randomPrivateKey();
    const publicKey = await ed25519.getPublicKey(secretKey);
    return new AetherKeypair(secretKey, publicKey);
  }

  /**
   * Create keypair from existing secret key
   */
  static async fromSecretKey(secretKey: Bytes): Promise<AetherKeypair> {
    const publicKey = await ed25519.getPublicKey(secretKey);
    return new AetherKeypair(secretKey, publicKey);
  }

  /**
   * Create keypair from seed phrase (deterministic)
   */
  static async fromSeed(seed: string): Promise<AetherKeypair> {
    const seedBytes = new TextEncoder().encode(seed);
    const secretKey = sha256(seedBytes);
    return AetherKeypair.fromSecretKey(secretKey);
  }

  /**
   * Sign a message
   */
  async sign(message: Bytes): Promise<Signature> {
    const signature = await ed25519.sign(message, this.secretKey);
    return Buffer.from(signature).toString('hex');
  }

  /**
   * Verify a signature
   */
  static async verify(signature: Signature, message: Bytes, publicKey: Bytes): Promise<boolean> {
    try {
      const sigBytes = Buffer.from(signature, 'hex');
      return await ed25519.verify(sigBytes, message, publicKey);
    } catch {
      return false;
    }
  }

  /**
   * Convert public key to Aether address
   */
  static publicKeyToAddress(publicKey: Bytes): Address {
    const hash = sha256(publicKey);
    // Take last 20 bytes of hash
    const addressBytes = hash.slice(-20);
    return '0x' + Buffer.from(addressBytes).toString('hex');
  }

  /**
   * Export secret key as hex string
   */
  toSecretKeyHex(): string {
    return Buffer.from(this.secretKey).toString('hex');
  }

  /**
   * Export public key as hex string
   */
  toPublicKeyHex(): string {
    return Buffer.from(this.publicKey).toString('hex');
  }

  /**
   * Create keypair from hex-encoded secret key
   */
  static async fromSecretKeyHex(hex: string): Promise<AetherKeypair> {
    const secretKey = Buffer.from(hex, 'hex');
    return AetherKeypair.fromSecretKey(secretKey);
  }
}

File: sdk/python/src/aether/keypair.py
"""
Keypair management for Aether

Ed25519 keypair generation and signing
"""

import hashlib
import nacl.signing
import nacl.encoding
from typing import Tuple
from .types import Address, Signature


class Keypair:
    """Ed25519 keypair for signing transactions"""

    def __init__(self, signing_key: nacl.signing.SigningKey):
        self._signing_key = signing_key
        self._verify_key = signing_key.verify_key
        self.public_key = bytes(self._verify_key)
        self.secret_key = bytes(signing_key)
        self.address = self.public_key_to_address(self.public_key)

    @classmethod
    def generate(cls) -> "Keypair":
        """Generate a new random keypair"""
        signing_key = nacl.signing.SigningKey.generate()
        return cls(signing_key)

    @classmethod
    def from_secret_key(cls, secret_key: bytes) -> "Keypair":
        """Create keypair from existing secret key"""
        signing_key = nacl.signing.SigningKey(secret_key)
        return cls(signing_key)

    @classmethod
    def from_seed(cls, seed: str) -> "Keypair":
        """Create keypair from seed phrase (deterministic)"""
        seed_bytes = seed.encode('utf-8')
        secret_key = hashlib.sha256(seed_bytes).digest()
        return cls.from_secret_key(secret_key)

    @classmethod
    def from_secret_key_hex(cls, hex_str: str) -> "Keypair":
        """Create keypair from hex-encoded secret key"""
        secret_key = bytes.fromhex(hex_str)
        return cls.from_secret_key(secret_key)

    def sign(self, message: bytes) -> Signature:
        """Sign a message"""
        signed = self._signing_key.sign(message)
        return signed.signature.hex()

    @staticmethod
    def verify(signature: Signature, message: bytes, public_key: bytes) -> bool:
        """Verify a signature"""
        try:
            verify_key = nacl.signing.VerifyKey(public_key)
            sig_bytes = bytes.fromhex(signature)
            verify_key.verify(message, sig_bytes)
            return True
        except:
            return False

    @staticmethod
    def public_key_to_address(public_key: bytes) -> Address:
        """Convert public key to Aether address"""
        hash_bytes = hashlib.sha256(public_key).digest()
        address_bytes = hash_bytes[-20:]
        return "0x" + address_bytes.hex()

    def to_secret_key_hex(self) -> str:
        """Export secret key as hex string"""
        return self.secret_key.hex()

    def to_public_key_hex(self) -> str:
        """Export public key as hex string"""
        return self.public_key.hex()

